### PR TITLE
Adding checks for methods in Random module to require array arguments to be rectangular

### PR DESCRIPTION
**Pull request by CaptainSharf**
As part of this PR, adding checks for methods in random module to take valid parameters. Specifically
- Adding `compileError` when the methods `fillRandom`,`shuffle` and `permutation` are invoked with non-rectangular arrays as arguments
- Adding a `compileError` message when the methods `shuffle` and `permutation` are invoked with `algorithm=RNG.NPB`
- Modifying the docs for each of these methods to reflect the above mentioned changes
- Adding tests for validating whether array arguments passed to the methods `fillRandom`,`shuffle` and `permutation` are not rectangular

This PR contains the possible fix for #16425.

### CODE REVIEW

**Comment by e-kayrakli** [comment]
I prefer the default rectangular checks also produce similar errors to the others. e.g. `fillRandom can only be used with rectangular arrays` or something.

With the implementation in the PR today, we'll just get resolution errors, which will not be clear when looking at the docs. Which reminds me that these limitations should be mentioned in the functions' docs.

**Comment by CaptainSharf** [comment]
Ok, would it be a good idea to throw compileErrors for each of these methods when the array is non rectangular?

Cool, I'll add a note in the documentation for these methods?

**Comment by e-kayrakli** [comment]
> Thanks for the feedback @e-kayrakli , would it be a good idea to throw compileErrors for each of these methods when the array is non rectangular?

I think so. We don't have a convention for handling such compile time errors, but at the very least a module should be consistent in how it reports errors. And a module reporting "I don't support using this argument for this function" is much better than the compiler saying "I can't resolve this function call".

> Cool, I'll add a note in the documentation for these methods?

Maybe not literally a `.. note::` but something in the description of arguments.

I'd also wait for someone more knowledgable in the Random module to chime in, in case they'd disagree with me. I think @mppf may have opinions here.

**Comment by bradcray** [comment]
> And a module reporting "I don't support using this argument for this function" is much better than the compiler saying "I can't resolve this function call".

I'd simply suggest pushing the rectangular check into the top of the routine, like:

```chapel
if (!isRectangularArr(arr)) then
  compilerError("...");
```

where a wording like Engin's looks good to me.  My rationale being: lack of where clauses simplifies function resolution which tends to be an expensive pass and handling it at the top of the function seems simple and clean in terms of handling cases and control flow.

But I'm also open to doing multiple overloads where the other is an error overload if anyone prefers that.

**Comment by bradcray** [comment]
Oh, and I think a test, or tests, should be added to lock this behavior in (somewhere under test/library/standard/Random/).  A _possibly_ useful tip to potentially have to avoid writing a number of tests is that I _think_ you can sometimes get reasonable behavior for a test that runs into multiple compilerErrors if you use the `--ignore-errors-for-pass` compiler flag.

File: modules/standard/Random.chpl [hunk 1]
```diff
@@ -141,15 +141,18 @@ module Random {
     compilerError("Random.fillRandom is only defined for numeric arrays");
   }
 
-  /* Shuffle the elements of an array into a random order.
+  /* Shuffle the elements of the array into a random order.
```
Comment (bradcray) [inline modules/standard/Random.chpl:144]: I think `an` is actually better here (since it could refer to any array, not a specific one).  Or `a rectangular`.

File: modules/standard/Random.chpl [hunk 2]
```diff
@@ -141,15 +141,18 @@ module Random {
...
-     :arg arr: a 1-D non-strided array
+     :arg arr: a rectangular 1-D non-strided array
      :arg seed: the seed to use when shuffling. Defaults to
       `oddCurrentTime` from :type:`RandomSupport.SeedGenerator`.
      :arg algorithm: A param indicating which algorithm to use. Defaults to PCG.
      :type algorithm: :type:`RNG`
    */
   proc shuffle(arr: [], seed: int(64) = SeedGenerator.oddCurrentTime, param algorithm=RNG.PCG) {
+    if(algorithm==RNG.NPB) then
+      compilerError("Cannot use NPB Random number generator for array shuffling");
```
Comment (bradcray) [inline modules/standard/Random.chpl:154]: What's the rationale for this change?  What about it doesn't work?

File: modules/standard/Random.chpl [hunk 3]
```diff
@@ -1041,7 +1046,10 @@ module Random {
         :arg arr: The array to be filled
         :type arr: [] :type:`eltType`
       */
-      proc fillRandom(arr: [] eltType) {
+      proc fillRandom(arr: [] eltType) throws {
```
Comment (bradcray) [inline modules/standard/Random.chpl:1049]: Does it throw?

File: modules/standard/Random.chpl [hunk 4]
```diff
@@ -1158,9 +1166,12 @@ module Random {
         return _choice(this, x, size=size, replace=replace, prob=prob);
       }
 
-      /* Randomly shuffle a 1-D array. */
+      /* Shuffle the elements of the array in random order. */
```
Comment (bradcray) [inline modules/standard/Random.chpl:1169]: Here too, I tend to prefer the original wording.

File: modules/standard/Random.chpl [hunk 5]
```diff
@@ -2627,7 +2642,10 @@ module Random {
         :arg arr: The array to be filled
         :type arr: [] :type:`eltType`
       */
-      proc fillRandom(arr: [] eltType) {
+      proc fillRandom(arr: [] eltType) throws {
```
Comment (bradcray) [inline modules/standard/Random.chpl:2645]: `throws` again

**Comment by bradcray** [comment]
A couple of other things:

* Your OP on this pull request should describe what the change does and why.  This will end up in the git history, so the goal is to document the change here in a way that is independent of the issue (because someone may clone this tree away from GitHub, years from now and lose the link to the issue)

* I think it's worth checking that `make docs` works and that the rendered documentation looks OK if you haven't (`make docs` is pretty fickle about spacing and such).

File: modules/standard/Random.chpl (hunk 2 above)
Comment (CaptainSharf) [reply modules/standard/Random.chpl:154]: The NPBRandom module does not implement the shuffle and permutation methods. So I thought it might be a good idea to throw compileError if shuffle/permutation is invoked with the NPB Random number generator.

File: modules/standard/Random.chpl (hunk 3 above)
Comment (CaptainSharf) [reply modules/standard/Random.chpl:1049]: sorry, I thought compileErrors are thrown. Will correct this

File: modules/standard/Random.chpl (hunk 2 above)
Comment (bradcray) [reply modules/standard/Random.chpl:154]: Huh...  So what is the role of the `param algorithm` argument if `.shuffle` is called as a method on a RandomStream?  It seems redundant (and looks like it's currently an unused argument?  It might be useful to have @mppf weigh in on this.

**Comment by CaptainSharf** [comment]
> Your OP on this pull request should describe what the change does and why.  This will end up in the git history, so the goal is to document the change here in a way that is independent of the issue (because someone may clone this tree away from GitHub, years from now and lose the link to the issue)

Please let me know if the current description is meaningful. Sorry, I don't know what OP stands for :P but I'm assuming it to be the description of the PR. Let me know otherwise

> I think it's worth checking that `make docs` works and that the rendered documentation looks OK if you haven't (`make docs` is pretty fickle about spacing and such).

Checked the docs, the doc page content seems to be formatted properly. Are there additional automation checks which I need to run for docs?

Thanks for your prompt feedback!!

**Comment by bradcray** [comment]
> Please let me know if the current description is meaningful

That's a big improvement, thanks.  When I have questions like I had last night ("Why is there this new check against algorithm = NPB?"), I refer to the OP (which stands for "original post" in online forums and github) to see if there's a rationale given there.  For most major changes, there should be (and for more complicated PRs, the OP often talks about the approach taken and such; for what you've done this is the right length).  And then this post goes into the permanent record when we merge, so it's also a good form of documentation when looking back on changes and why they occurred.

> Checked the docs, the doc page content seems to be formatted properly. Are there additional automation checks which I need to run for docs?

Nope, that should be good enough, thanks.  When we're done reviewing and ready to merge, we'll also run a full run of the test suite against the diff on our side to make sure there are no surprises.

File: modules/standard/Random.chpl (hunk 2 above)
Comment (bradcray) [reply modules/standard/Random.chpl:155]: (specifically, I'm wondering what the behavior of `myPCGRandomStream.shuffle(algorithm=RNG.NPB)` or `myNPBStream.shuffle(algorithm=RNG.PCG)` would be)

Comment (CaptainSharf) [reply modules/standard/Random.chpl:155]: Hi @bradcray,
The implementation of permutation/shuffle in the individual modules `PCGRandom` and `NPBRandom` do not accept algorithm as a parameter. In other words what I mean to say is the methods shuffle/permutation in the Random module provide a wrapper for the corresponding methods in the individual modules `PCGRandom` and `NPBRandom`

Comment (bradcray) [reply modules/standard/Random.chpl:155]: Ah, I think I understand better now.  I was thinking this was a method on either PCGRandom or NPBRandom, but it's actually a wrapper function that dispatches to a stream of one type or the other.  And the NPB stream doesn't support the shuffle() routine.  Thanks for clarifying.

Comment (bradcray) [reply modules/standard/Random.chpl:155]: A better error message for cases like this might be "The NPB random number generator currently does not support array shuffling" since it could be extended to do so in the future?  (not that that's a particularly high priority).

**Comment by bradcray** [comment]
Seeing the merge for this arrive made me realize that we don't have an ICLA on-file for @CaptainSharf, as the bot indicated on this PR.  I've reached out by gitter to try and address this retroactively.
